Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,289 @@
+/-
+Copyright (c) 2025 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson, Jun Kwon
+-/
+import Mathlib.Data.Set.Basic
+
+/-!
+# Multigraphs
+
+A multigraph is a set of vertices and a set of edges,
+together with incidence data that associates each edge `e`
+with an unordered pair `(x,y)` of vertices called the 'ends' of `e`.
+`x` and `y` may be equal, in which case `e` is a 'loop',
+and there may also be more than one edge `e` whose ends are the same pair `(x,y)`.
+A multigraph where neither of these occurs is 'simple',
+and these objects are described by `SimpleGraph`.
+
+This module defines `Graph α β` for a vertex type `α` and an edge type `β`,
+and gives basic API for incidence, adjacency and extensionality.
+The design broadly follows [Chou1994].
+
+## Main definitions
+
+For `G : Graph α β`, ...
+
+* `G.V` denotes the vertex set of `G` as a term in `Set α`.
+* `G.E` denotes the edge set of `G` as a term in `Set β`.
+* `G.Inc₂ e x y` means that the edge `e : β` has vertices `x : α` and `y : α` as its ends.
+* `G.Inc e x` means that the edge `e : β` has `x` as one of its ends.
+* `G.Adj x y` means that there is an edge `e` having `x` and `y` as its ends.
+* `G.IsLoopAt e x` means that `e` is a loop edge with both ends equal to `x`.
+* `G.IsNonloopAt e x` means that `e` is a non-loop edge with one end equal to `x`.
+
+## Implementation notes
+
+Unlike the design of `SimpleGraph`, the vertex and edge sets of `G` are modelled as sets
+`G.V : Set α` and `G.E : Set β`, within ambient types, rather than being types themselves.
+This mimics the 'embedded set' design used in `Matroid`, which seems to be more convenient for
+formalizing real-world proofs in combinatorics.
+
+A specific advantage is that this will allow subgraphs of `G : Graph α β` to also exist on
+an equal footing with `G` as terms in `Graph α β`,
+and so there is no need for an extensive `Graph.subgraph` API and all the associated
+definitions and canonical coercion maps. The same will go for minors and the various other
+partial orders on multigraphs.
+
+The main tradeoff is that parts of the API will require caring about whether a term
+`x : α` or `e : β` is a 'real' vertex or edge of the graph, rather than something outside
+the vertex or edge set. This is an issue, but is likely amenable to automation.
+
+## Notation
+
+Reflecting written mathematics, we use the compact notations `G.V` and `G.E` to describe
+vertex and edge sets formally, but use the longer terms `vxSet` and `edgeSet` within
+lemma names to refer to the same objects.
+
+-/
+
+variable {α β : Type*} {x y z u v w : α} {e f : β}
+
+open Set
+
+/-- A multigraph with vertex set `V : Set α` and edge set `E : Set β`,
+as described by a predicate describing whether an edge `e : β` has ends `x` and `y`. -/
+structure Graph (α β : Type*) where
+  /-- The vertex set. -/
+  V : Set α
+  /-- The edge set. -/
+  E : Set β
```

Comment:
I can see the argument for `V`, but why do we need `E`? Can't this be encoded into `Inc₂`?

Reply:
It can, but I've included it for good definitional properties. The extensionality section has an alternative constructor that doesn't need `E`.

Reply:
I'm not fully convinced; why do we care about the defeq of `E` but not `Inc`?

Usually we only provide fields for "good definitional properties" when the alternative is instance diamonds / DTT hell, which as far as I can tell is not relevant here.

At any rate, if you insist on keeping the field for this reason, please mention that it exists only for nice defeqs in the docstring (referencing `edgeSet_eq_setOf_exists_inc₂` which shows it is fixed)

Reply:
The first place this comes up is when you restrict to a set of edges. It's convenient if `(G.edgeRestrict F).E = F` is `rfl`. 

I'll add an explanation in the docstring.

Reply:
Edge-restriction is a little weird, because in fact one needs `(G.edgeRestrict F).E = F ∩  G.E`, though this is still a convenient defeq. 

In my repo, I have used the defeq `rfl : (G.edgeDelete F).E = G.E \ F` quite a few times, and something similar will work for contraction.